README.md
[![Gem Version](https://badge.fury.io/rb/kuby-anycable.svg)](https://rubygems.org/gems/kuby-anycable)

# Kuby AnyCable

[Kuby][] plugin to deploy [AnyCable][] applications.

This plugin allows you to install all the required AnyCable components to a Kubernetes cluster.

## Installation

Add to your project:

```ruby
# Gemfile
gem "kuby-anycable"
```

## Usage

Here is the minimal configuration:

```ruby
# kuby.rb

require "kuby-anycable"

# ...

Kuby.define("my-app") do
  environment(:production) do
    #...

    kubernetes do
      add_plugin :rails_app do
        # ...
      end

      add_plugin :anycable_rpc
      add_plugin :anycable_go
    end
  end
end
```

What happens under the hood:

- RPC service definitions are created:
  - The current `rails_app` image is used for the container.
  - The `rails_app` Config Map is attached to the RPC container.
- AnyCable-Go service definitions are created:
  - The latest stable image is used for `anycable-go`.
  - Connected to the RPC service (using [DNS load balancing](https://docs.anycable.io/deployment/load_balancing?id=client-side-load-balancing) by default).
  - Redis URL is inferred from the RPC service `ANYCABLE_REDIS_URL` or `REDIS_URL`.
  - [Concurrency settings](https://docs.anycable.io/anycable-go/configuration?id=concurrency-settings) are adjusted according to the number of RPC servers and their concurrency settings.

Of course, you can customize the resources:

```ruby
add_plugin :anycable_rpc do
  replicas 2
  # Provide Redis URL explicitly.
  redis_url "redis://custom_url"
  # Override gRPC server port (but why?)
  port 50051
  # Expose additional port named 'metrics'
  # (e.g., if you use Prometheus exporter).
  # Disabled by default.
  metrics_port 3030
  # Shortcut for ENV['ANYCABLE_RPC_SERVER_ARGS__MAX_CONNECTION_AGE_MS']
  max_connection_age 300000
  # Shortcut for ENV['ANYCABLE_RPC_POOL_SIZE']
  rpc_pool_size 30
end

add_plugin :anycable_go do
  replicas 2
  # Provide Redis URL explicitly.
  redis_url "redis://custom_url"
  # Override web server port (but why?)
  port 8081
  # Metrics port (enabled by default and exposed as "metrics")
  metric_port 5001
  # Provide path to RPC server explicitly
  rpc_host "my-app-rpc:50051"
  # WebSocket endpoint path
  ws_path "/cable"
  # AnyCable-Go Dockerimage
  image "anycable/anycable-go:1.1"
  # Specify ENV['ANYCABLE_RPC_CONCURRENCY'] explicitly
  rpc_concurrency nil
  # Use a separate hostname for AnyCable-Go
  # (disabled by default)
  hostname nil
end
```

### Using with Alpine images

Installing Ruby deps on Apline images requires some special attention to gRPC-related gems (tl;dl we need to build them from source).

Kuby AnyCable comes with a special _package_, which installs everything for you (so you shouldn't use any hacks yourself). You need to add a single line:

```ruby
Kuby.define("my-app") do
  environment(:production) do
    docker do
      # ...
      distro :apline
      package_phase.add("anycable-build")

      # ...
    end

    # ...
  end
end
```

## Contributing

Bug reports and pull requests are welcome on GitHub at [https://github.com/palkan/kuby-anycable](https://github.com/palkan/kuby-anycable).

## License

The gem is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).

[Kuby]: https://getkuby.io
[AnyCable]: https://anycable.io


CHANGELOG.md
# Change log

## master

## 0.1.0 (2021-11-24)

- Initial.

[@palkan]: https://github.com/palkan


lib/kuby/anycable/rpc_plugin.rb
# frozen_string_literal: true

module Kuby
  module AnyCable
    class RPCPlugin < ::Kuby::Plugin
      extend ::KubeDSL::ValueFields

      ROLE = "rpc"

      value_fields :replicas
      value_fields :redis_url
      value_fields :port
      value_fields :metrics_port
      value_fields :max_connection_age
      value_fields :rpc_pool_size

      def after_initialize
        @replicas = 1
        @port = 50051
        @metrics_port = nil
        @max_connection_age = 300000
        @rpc_pool_size = 30
      end

      def after_configuration
        return unless rails_spec

        deployment.spec.template.spec.container(:rpc).merge!(
          rails_spec.deployment.spec.template.spec.container(:web), fields: [:env_from]
        )
      end

      def configure(&block)
        instance_eval(&block) if block
      end

      def before_deploy(manifest)
        image_with_tag = "#{docker.image.image_url}:#{kubernetes.tag || Kuby::Docker::LATEST_TAG}"

        deployment do
          spec do
            template do
              spec do
                container(:rpc) do
                  image image_with_tag
                end
              end
            end
          end
        end
      end

      def resources
        @resources ||= [
          service_account,
          service,
          config_map,
          deployment
        ]
      end

      def service_account(&block)
        context = self

        @service_account ||= KubeDSL.service_account do
          metadata do
            name "#{context.selector_app}-#{ROLE}-sa"
            namespace context.namespace.metadata.name

            labels do
              add :app, context.selector_app
              add :role, ROLE
            end
          end
        end

        @service_account.instance_eval(&block) if block
        @service_account
      end

      def service(&block)
        spec = self

        @service ||= KubeDSL.service do
          metadata do
            name "#{spec.selector_app}-#{ROLE}-svc"
            namespace spec.namespace.metadata.name

            labels do
              add :app, spec.selector_app
              add :role, ROLE
            end
          end

          spec do
            type "ClusterIP"
            cluster_ip "None"

            selector do
              add :app, spec.selector_app
              add :role, ROLE
            end

            port do
              name "rpc"
              port spec.port
              protocol "TCP"
              target_port "rpc"
            end

            if spec.metrics_port
              port do
                name "metrics"
                port spec.metrics_port
                protocol "TCP"
                target_port "metrics"
              end
            end
          end
        end

        @service.instance_eval(&block) if block
        @service
      end

      def deployment(&block)
        context = self

        @deployment ||= KubeDSL.deployment do
          metadata do
            name "#{context.selector_app}-#{ROLE}"
            namespace context.namespace.metadata.name

            labels do
              add :app, context.selector_app
              add :role, ROLE
            end
          end

          spec do
            replicas context.replicas

            selector do
              match_labels do
                add :app, context.selector_app
                add :role, ROLE
              end
            end

            strategy do
              type "RollingUpdate"

              rolling_update do
                max_surge "25%"
                max_unavailable 0
              end
            end

            template do
              metadata do
                labels do
                  add :app, context.selector_app
                  add :role, ROLE
                end
              end

              spec do
                container(:rpc) do
                  name "#{context.selector_app}-#{ROLE}"
                  image_pull_policy "IfNotPresent"
                  command %w[bundle exec anycable]

                  port do
                    container_port context.port
                    name "grpc"
                    protocol "TCP"
                  end

                  if context.metrics_port
                    port do
                      container_port context.metrics_port
                      name "metrics"
                      protocol "TCP"
                    end
                  end

                  env_from do
                    config_map_ref do
                      name context.config_map.metadata.name
                    end
                  end

                  readiness_probe do
                    success_threshold 1
                    failure_threshold 3
                    initial_delay_seconds 15
                    period_seconds 10
                    timeout_seconds 3

                    tcp_socket do
                      port "grpc"
                    end
                  end

                  liveness_probe do
                    success_threshold 1
                    failure_threshold 3
                    initial_delay_seconds 90
                    period_seconds 10
                    timeout_seconds 3

                    tcp_socket do
                      port "grpc"
                    end
                  end
                end

                image_pull_secret do
                  name context.kubernetes.registry_secret.metadata.name
                end

                restart_policy "Always"
                service_account_name context.service_account.metadata.name
              end
            end
          end
        end

        @deployment.instance_eval(&block) if block
        @deployment
      end

      def config_map(&block)
        spec = self

        @config_map ||= KubeDSL.config_map do
          metadata do
            name "#{spec.selector_app}-#{ROLE}-config"
            namespace spec.namespace.metadata.name
          end

          data do
            add "ANYCABLE_RPC_HOST", "0.0.0.0:50051"
            if spec.redis_url
              add "ANYCABLE_REDIS_URL", spec.redis_url
            end
            if spec.max_connection_age
              add "ANYCABLE_RPC_SERVER_ARGS__MAX_CONNECTION_AGE_MS", spec.max_connection_age.to_s
            end

            if spec.rpc_pool_size
              add "ANYCABLE_RPC_POOL_SIZE", spec.rpc_pool_size.to_s
            end
          end
        end

        @config_map.instance_eval(&block) if block
        @config_map
      end

      alias_method :env, :config_map

      delegate :kubernetes, to: :environment

      delegate :docker, to: :environment

      delegate :selector_app, to: :kubernetes

      delegate :namespace, to: :kubernetes

      def rails_spec
        @rails_spec ||= kubernetes.plugin(:rails_app)
      end
    end
  end
end


lib/kuby/anycable/go_plugin.rb
# frozen_string_literal: true

module Kuby
  module AnyCable
    class GoPlugin < ::Kuby::Plugin
      extend ::KubeDSL::ValueFields

      ROLE = "ws"

      value_fields :replicas
      value_fields :redis_url
      value_fields :port
      value_fields :metrics_port
      value_fields :rpc_host
      value_fields :rpc_concurrency
      value_fields :ws_path
      value_fields :image
      value_fields :hostname

      DEFAULT_IMAGE = "anycable/anycable-go:1.1"

      def after_initialize
        @replicas = 1
        @metrics_port = 5001
        @rpc_host = nil
        @rpc_concurrency = nil
        @redis_url = nil
        @ws_path = "/cable"
        @port = 8081
        @hostname = nil
        @image = DEFAULT_IMAGE
      end

      def after_configuration
        configure_rpc_host
        configure_redis_url
        configure_concurrency

        if rails_spec
          @hostname ||= rails_spec.hostname
          configure_ingress(rails_spec.ingress, hostname)
        end
      end

      def configure(&block)
        instance_eval(&block) if block
      end

      def resources
        @resources ||= [
          service_account,
          service,
          config_map,
          deployment
        ]
      end

      def configure_ingress(ingress, hostname)
        spec = self

        ingress.spec.rule do
          host hostname

          http do
            path do
              path spec.ws_path

              backend do
                service_name spec.service.metadata.name
                service_port spec.service.spec.ports.first.port
              end
            end
          end
        end
      end

      def service_account(&block)
        context = self

        @service_account ||= KubeDSL.service_account do
          metadata do
            name "#{context.selector_app}-#{ROLE}-sa"
            namespace context.namespace.metadata.name

            labels do
              add :app, context.selector_app
              add :role, ROLE
            end
          end
        end

        @service_account.instance_eval(&block) if block
        @service_account
      end

      def service(&block)
        spec = self

        @service ||= KubeDSL.service do
          metadata do
            name "#{spec.selector_app}-#{ROLE}-svc"
            namespace spec.namespace.metadata.name

            labels do
              add :app, spec.selector_app
              add :role, ROLE
            end
          end

          spec do
            type "ClusterIP"

            selector do
              add :app, spec.selector_app
              add :role, ROLE
            end

            port do
              name "http"
              port spec.port
              protocol "TCP"
              target_port "http"
            end

            if spec.metrics_port
              port do
                name "metrics"
                port spec.metrics_port
                protocol "TCP"
                target_port "metrics"
              end
            end
          end
        end

        @service.instance_eval(&block) if block
        @service
      end

      def deployment(&block)
        context = self

        @deployment ||= KubeDSL.deployment do
          metadata do
            name "#{context.selector_app}-#{ROLE}"
            namespace context.namespace.metadata.name

            labels do
              add :app, context.selector_app
              add :role, ROLE
            end
          end

          spec do
            replicas context.replicas

            selector do
              match_labels do
                add :app, context.selector_app
                add :role, ROLE
              end
            end

            strategy do
              type "RollingUpdate"

              rolling_update do
                max_surge "25%"
                max_unavailable 0
              end
            end

            template do
              metadata do
                labels do
                  add :app, context.selector_app
                  add :role, ROLE
                end
              end

              spec do
                container(:ws) do
                  name "#{context.selector_app}-#{ROLE}"
                  image context.image
                  image_pull_policy "IfNotPresent"

                  port do
                    container_port context.port
                    name "http"
                    protocol "TCP"
                  end

                  if context.metrics_port
                    port do
                      container_port context.metrics_port
                      name "metrics"
                      protocol "TCP"
                    end
                  end

                  env_from do
                    config_map_ref do
                      name context.config_map.metadata.name
                    end
                  end

                  readiness_probe do
                    success_threshold 1
                    failure_threshold 3
                    initial_delay_seconds 15
                    period_seconds 10
                    timeout_seconds 3

                    tcp_socket do
                      port "http"
                    end
                  end

                  liveness_probe do
                    success_threshold 1
                    failure_threshold 3
                    initial_delay_seconds 90
                    period_seconds 10
                    timeout_seconds 3

                    tcp_socket do
                      port "http"
                    end
                  end
                end

                image_pull_secret do
                  name context.kubernetes.registry_secret.metadata.name
                end

                restart_policy "Always"
                service_account_name context.service_account.metadata.name
              end
            end
          end
        end

        @deployment.instance_eval(&block) if block
        @deployment
      end

      def config_map(&block)
        spec = self

        @config_map ||= KubeDSL.config_map do
          metadata do
            name "#{spec.selector_app}-#{ROLE}-config"
            namespace spec.namespace.metadata.name
          end

          data do
            if spec.rpc_host
              add "ANYCABLE_RPC_HOST", spec.rpc_host
            end

            if spec.redis_url
              add "ANYCABLE_REDIS_URL", spec.redis_url
            end

            add "ANYCABLE_HOST", "0.0.0.0"
            add "ANYCABLE_PORT", spec.port.to_s

            if spec.metrics_port
              add "ANYCABLE_METRICS_PORT", spec.metrics_port.to_s
              add "ANYCABLE_METRICS_HTTP", "/metrics"
            end

            if spec.rpc_concurrency
              add "ANYCABLE_RPC_CONCURRENCY", spec.rpc_concurrency.to_s
            end
          end
        end

        @config_map.instance_eval(&block) if block
        @config_map
      end

      alias_method :env, :config_map

      delegate :kubernetes, to: :environment

      delegate :docker, to: :environment

      delegate :selector_app, to: :kubernetes

      delegate :namespace, to: :kubernetes

      def rpc_spec
        @rpc_spec ||= kubernetes.plugin(:anycable_rpc)
      end

      def rails_spec
        @rails_spec ||= kubernetes.plugin(:rails_app)
      end

      def configure_rpc_host
        return if config_map.data.get("ANYCABLE_RPC_HOST")
        # Make it possible to avoid setting RPC host at all
        return if rpc_host == false

        return unless rpc_spec

        config_map.data.add("ANYCABLE_RPC_HOST", "dns:///#{rpc_spec.service.metadata.name}:50051")
      end

      def configure_redis_url
        return if config_map.data.get("REDIS_URL") || config_map.data.get("ANYCABLE_REDIS_URL")
        # Make it possible to avoid setting Redis url at all
        return if redis_url == false

        # Try to lookup Redis url from the RPC and Web app specs
        [rpc_spec, rails_spec].compact.detect do |spec|
          %w[ANYCABLE_REDIS_URL REDIS_URL].detect do |env_key|
            url = spec.config_map.data.get(env_key)
            next unless url

            config_map.data.add("ANYCABLE_REDIS_URL", url)
            true
          end
        end
      end

      def configure_concurrency
        return if config_map.data.get("ANYCABLE_RPC_CONCURRENCY")
        return unless rpc_spec

        rpc_pool_size = rpc_spec.config_map.data.get("ANYCABLE_RPC_POOL_SIZE")
        return unless rpc_pool_size

        rpc_replicas = rpc_spec.replicas

        concurrency = ((rpc_pool_size.to_i * rpc_replicas) * 0.95 / replicas).round

        config_map.data.add("ANYCABLE_RPC_CONCURRENCY", concurrency.to_s)
      end
    end
  end
end
